aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -91,5 +95,49 @@ Tensor spdiags(
   return result_coo;
 }
 
+Tensor spdiags_backward(
+    const Tensor& grad_out,
+    const Tensor& offsets,
+    IntArrayRef input_shape) {
+  auto offsets_1d = offsets.dim() == 0 ? offsets.unsqueeze(0) : offsets;
+
+  auto n_diag = input_shape.size() == 2 ? input_shape[0] : 1;
+  auto n_col_in = input_shape.size() == 2 ? input_shape[1] : input_shape[0];
+  AT_ASSERT(grad_out.dim() == 2);
+  AT_ASSERT(offsets_1d.size(0) == n_diag);
+  auto grad_in_options = grad_out.options().layout(Layout::Strided);
+  // zeros since we are only going to se the non-zero elements
+  Tensor grad_in = at::zeros({n_diag, n_col_in}, grad_in_options);
+  auto grad_out_coo = grad_out.layout() == Layout::Sparse
+      ? grad_out.coalesce()
+      : grad_out.to_sparse();
+
+  auto grad_out_values = sparse::get_sparse_impl(grad_out_coo)->values_;
+  auto grad_out_indices = sparse::get_sparse_impl(grad_out_coo)->indices_;
+  auto grad_out_row_indices = grad_out_indices[0];
+  auto grad_out_col_indices = grad_out_indices[1];
+
+  // Precompute input row indices for each nnz
+  auto row_in_indices =
+      offsets_1d
+          .eq(grad_out_col_indices.sub(grad_out_row_indices).reshape({-1, 1}))
+          .nonzero()
+          .permute({1, 0})[-1];
```

Comment:
Why do you make it a 2D tensor again?

Reply:
I think this could be simplified to `.eq(grad_out_col_indices - grad_out_row_indices).nonzero().squeeze(-1)`.

Reply:
`grad_out_col_indices` and `grad_out_row_indices` are 1D tensors, so `nonzero` will give you the `(nnz, 1)` shape.

Reply:
`grad_out_col_indices` and `grad_out_row_indices` are 1D tensors already, so `nonzero` is going to give you the `(nnz, 1)` shape.

Reply:
Too tired... Sorry for the noise. Right, so, `grad_out_col_indices` and `grad_out_row_indices` are 1D already, so `nonzero` will give you a `(nnz, 1)` shape.

Reply:
Right, sorry, missed the order. Are these the same offsets from the forward?

Reply:
Nevermind, offsets_1d has a different shape from the indices.